# Phase 8: Intelligent Alert System Validation
**SmartGrid Sentinel Journal Experiment Series - Notebook 07**

## 1. Alert Validation Objective
SmartGrid Sentinel is an integrated early warning platform. The system does not merely produce raw numerical predictions; it translates model sequence outputs into operational risk alert levels, priority tiers, natural language explanations, and actionable protocols.
This notebook presents the empirical verification of the intelligent alert system, verifying five core operational alert requirements across risk levels, confidence thresholds, and invalid input exception handling.

## 2. Alert System Architecture & Decision Flow

```
Model Sequence Prediction (Informer 2H Horizon)
       ↓
Softmax Probabilities [P_High, P_Low, P_Medium]
       ↓
Risk Level Classification (High / Medium / Low)
       ↓
Alert Rules & Physical Telemetry Validation
       ↓
Alert Priority Level (RED / YELLOW / GREEN)
       ↓
Natural Language Explanation & Action Recommendations
```

In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use('seaborn-v0_8-whitegrid' if 'seaborn-v0_8-whitegrid' in plt.style.available else 'default')
plt.rcParams['font.sans-serif'] = 'DejaVu Sans'
plt.rcParams['axes.edgecolor'] = '#cccccc'
plt.rcParams['axes.linewidth'] = 0.8

alert_dir = '../alert_validation_phase8'
if not os.path.exists(alert_dir):
    alert_dir = 'alert_validation_phase8'

print('Environment initialized.')

Environment initialized.


## 3. Master Alert Validation Summary

Loading empirical test verification results across all five operational alert test cases from `alert_system_validation_results.csv`:

In [2]:
alert_df = pd.read_csv(os.path.join(alert_dir, 'alert_system_validation_results.csv'))
print('=== ALERT SYSTEM VALIDATION SUMMARY TABLE ===')
print(alert_df.to_string(index=False))

=== ALERT SYSTEM VALIDATION SUMMARY TABLE ===
                                      Test Case             Input Risk Level                           Alert Priority Correct Explanation Generated                            Correct Action Protocol Status
                        Low-Risk Alert Behavior                          Low                 GREEN (Normal Operation)                          True                                               True PASSED
                     Medium-Risk Alert Behavior                       Medium                YELLOW (Warning Advisory)                          True                                               True PASSED
                       High-Risk Alert Behavior                         High                   RED (Urgent Emergency)                          True                                               True PASSED
            Invalid/Missing Input Safe Handling Invalid Upazila / Empty JSON          HTTP 400 / HTTP 422 Error Catch             

## 4. Behavior Breakdown by Risk & Exception Condition

### A. Low-Risk Alert Behavior (GREEN Priority)
- **Trigger Condition:** Model predicts `Low` risk (e.g., Transformer Load 35%, Demand 85 MW).
- **Mapped Priority:** `GREEN (Normal Operation)`.
- **Generated Explanation:** *"The grid is currently operating under stable conditions. The transformer is operating at 35.0% load, which is currently within a comfortable range."*
- **Action Protocol:** `✓ Normal electricity usage can continue. ✓ No special precautions are needed.`

### B. Medium-Risk Alert Behavior (YELLOW Priority)
- **Trigger Condition:** Model predicts `Medium` risk (e.g., Transformer Load 68.5%, Demand 165 MW).
- **Mapped Priority:** `YELLOW (Warning Advisory)`.
- **Generated Explanation:** *"The grid is currently under moderate stress. The transformer is operating at 68.5% load, which indicates moderate loading."*
- **Action Protocol:** `• Avoid unnecessary high-power appliance usage during peak hours. • Charge essential devices.`

### C. High-Risk Alert Behavior (RED Priority)
- **Trigger Condition:** Model predicts `High` risk (e.g., Transformer Load 92.4%, Demand 245 MW).
- **Mapped Priority:** `RED (Urgent Emergency)`.
- **Generated Explanation:** *"The grid is showing signs of significant stress. The transformer is heavily loaded at 92.4%, so continued monitoring is important."*
- **Action Protocol:** `⚠ Charge essential devices immediately. ⚠ Limit non-essential electricity consumption.`

## 5 & 6. Confidence Behavior & Invalid Input Handling

### D. Confidence Threshold & Softmax Behavior
- Validates that Softmax output probabilities $[P_{\text{High}}, P_{\text{Low}}, P_{\text{Medium}}]$ strictly sum to $1.0000$.
- Verified live API confidence evaluation (Top probability `63.92%` for `Medium` risk prediction).

### E. Invalid & Missing Input Exception Handling
- Verifies safe HTTP 400 and HTTP 422 exception catches for unknown Upazilas or malformed JSON payloads.
- Confirms that backend server operations remain stable without process crashes.

In [3]:
# Display Test Status Count
status_counts = alert_df['Status'].value_counts()
print('=== ALERT SYSTEM TEST STATUS SUMMARY ===')
for status, count in status_counts.items():
    print(f'Status [{status}]: {count} / {len(alert_df)} Test Cases')

=== ALERT SYSTEM TEST STATUS SUMMARY ===
Status [PASSED]: 5 / 5 Test Cases


## 7. Journal-Ready Discussion & Conclusions

- **Operational Utility:** The alert engine establishes a operational bridge between time-series ML prediction models and real-world grid operator decision workflows.
- **Fault Tolerance:** Automated exception handling ensures high platform availability when handling missing telemetry mappings or malformed payload requests.